node/tests/swamp/swamp.go

```diff
@@ -0,0 +1,169 @@
+package swamp
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/stretchr/testify/require"
+
+	"github.com/celestiaorg/celestia-node/core"
+	"github.com/celestiaorg/celestia-node/libs/keystore"
+	"github.com/celestiaorg/celestia-node/node"
+	"github.com/celestiaorg/celestia-node/node/p2p"
+)
+
+var blackholeIP6 = net.ParseIP("100::")
+
+// TODO(@Bidon15): Issue #350
+
+// Swamp represents the main functionality that is needed for the test-case:
+// - Network to link the nodes
+// - CoreClient to share between Bridge nodes
+// - Slices of created Bridge Nodes / Light Clients
+// - trustedHash taken from the CoreClient and shared between nodes
+type Swamp struct {
+	t            *testing.T
+	Network      mocknet.Mocknet
+	CoreClient   core.Client
+	BridgeNodes  []*node.Node
+	LightClients []*node.Node
+	trustedHash  string
+}
+
+// NewSwamp creates a new instance of Swamp.
+func NewSwamp(t *testing.T) *Swamp {
+	if testing.Verbose() {
+		log.SetDebugLogging()
+	}
+
+	ctx := context.Background()
+	// TODO(@Bidon15): Rework this to be configurable by the swamp's test
+	// case, not here.
+	kvStore := core.CreateKvStore(200)
+	coreNode := core.StartMockNode(kvStore)
+
+	coreNode.Config().Consensus.CreateEmptyBlocksInterval = 200 * time.Millisecond
+
+	core := core.NewEmbeddedFromNode(coreNode)
+
+	swp := &Swamp{t: t, Network: mocknet.New(ctx), CoreClient: core}
+	swp.trustedHash = swp.getTrustedHash(ctx)
+
+	return swp
+}
+
+// WaitTillHeight holds the test execution until the given amount of blocks
+// have been produced by the CoreClient.
+func (s *Swamp) WaitTillHeight(height int64) {
```

Comment:
Blocking operations should rcv a context on them. Then pass to both subscribe and unsubscribe. Lastly, make a select inside for with ctx and <-blocks